lib/service/tracking/tracking_cubit.dart

```diff
@@ -0,0 +1,18 @@
+import 'package:flutter_bloc/flutter_bloc.dart';
+import 'package:tracking_repository/tracking_repository.dart';
+
+class _TrackingState {
+  const _TrackingState();
+}
+
+class TrackingCubit extends Cubit<_TrackingState> {
+  TrackingCubit(this.trackingRepository) : super(const _TrackingState());
+
+  final TrackingRepository trackingRepository;  
+}
+
+extension LandingPageTracking on TrackingCubit { 
+  void onPressedStartTrading() {
+    trackingRepository.track(LandingPageEvent.onPressedStartTrading());
+  }
+}
```

Comment:
The cubit itself is meant to hold methods that you'll be calling.
Please move it inside cubit and make it reusable, not as specific as this one is.
A name like `trackEvent` would be great, same goes for the repository method, `track` is a little bit too generic.

Reply:
I am not sure I understand all you ask.

Having extensions on that class only increases readability and offers meaning to the usage of these methods. Maybe I don't know of some hidden side effect.

I would advise against having to pass directly names & params from UI / blocs directly to the repository only to pass through this bloc.
This will affect in a negative way encapsulation, scalability and readability down the line.

However, I can update as you guys want, I am not "attached" to this solution.

Reply:
It doesn't make sense to create extensions on a class that it's sole purpose is to allow you to declare methods for your business logic.

Would you mind giving me an e2e example of what the event would look like and how you'd call it from the UI? maybe I'm missing something.

Reply:
About the usage of extensions:
Instead of having N methods in a huge file you would get smaller more explicit files / 'feature'.
I do agree that extensions may not be the best instrument to achieve the goal I had in mind and I am opened to suggestions.


About defining each event as opposed to one unified "track(name: params:)" at the BL layer.
One immediate example of a benefit would be that this increases readability. All events that must be tracked are consolidated in that feature (event A/B/C/D ....) and not have to search & hunt these in the codebase. 
A good impact is also on the robustness when rework/refactoring is done a feature, it is clearer if you missed to add any defined event after cleaning / clearing out old code from a feature.

Here are a few more usage example and how they would differ


`
/// UI KISS
  context.read<TrackingCubit>()..trackEvent({
      'name': 'pressed_button_delete',
      'params': {
        'event_type': 'button_action',
        'user_type': 'admin',      
        'user_id': 100,
        'reason': 'special_limitation',
        'additional_information': 12,
      },
    }
  );  

  context.read<TrackingCubit>()..trackEvent({
      'name': 'admin_panel_screen',
      'params': {
        'event_type': 'screen_view',
        'user_type': 'admin',      
      }
  );  

  context.read<TrackingCubit>()..trackEvent({
      'name': 'pressed_button_register',
      'params' : {
        'event_type': 'button_action',
        'user_type': 'control',      
        'status': 'succesful'
      }
    }
  );  

  context.read<TrackingCubit>()..trackEvent({
      'name': 'pressed_button_register',
      'params' : {
        'event_type': 'button_action',
        'user_type': 'control',      
        'status': 'failed',
        'reason': 'weak_password',
      }
    }
  );
  

  // UI encapsulated

  context.read<TrackingCubit>()..onPressedDeleteUser(    
    userId: 100,
    reason: DeleteReason.specialLimitation,
    aditionalInformation: 12,
  );

  context.read<TrackingCubit>()..openedAdminPanel();

  context.read<TrackingCubit>()..onSuccesfulPressedRegister();

  context.read<TrackingCubit>()..onFailedPressedRegister(
    reason: FailedRegister.weakPassword,
  );  


// A lot of meaningless info for the UI layer is handled at the BL
// definitions of Keys can be kept separated from the UI layer
// Additional transformations/ data preparation that might occur as well
                             
abstract class TrackEvent {
  TrackEvent({ 
     required this.name,
     Map<String, dynamic>? params,
  }) : parameters = params ?? {};

  /// name of the event
  final String name;
  
  final Map<String, dynamic> parameters;
}

class AdminPanelEvents extends TrackEvent { 
  AdminPanelEvents.onPressedDeleteUser({
    required int userId,
    required DeleteReason reason,
    required int aditionalInformation,
  }) : super (
    name: 'pressed_button_delete',
    params: {
      'event_type': UtilsE(EventType.buttonAction), // or screen_view
      'user_role': UtilsU(UserRole.admin),
      'user_id': userId,
      'reason': reason,
      'additional_information': aditionalInformation,      
    }
  );
  
  AdminPanelEvents.openedAdminPanel({
    required int userId,
    required DeleteReason reason,
    required int aditionalInformation,
  }) : super (
    name: 'admin_panel_screen',
    params: {
      'event_type': UtilsE(EventType.screenView), // or screen_view
      'user_role': UtilsU(UserRole.admin),      
    }
  );
}

class RegistrationEvents extends TrackEvent {
   RegistrationEvents.onSuccesfulPressedRegister() : super(
      name: 'pressed_button_register',
      params : {
        'event_type': UtilsE(EventType.buttonAction),
        'user_type': UtilsU(UserRole.control),      
        'status': 'succesful',
      }    
  );

  RegistrationEvents.onFailedPressedRegister(
    FailedRegister reason,
  ): super(
      name: 'pressed_button_register',
      params : {
        'event_type': e(EventType.buttonAction),
        'user_type': u(UserRole.control),      
        'status': 'failed',
        'reason': f(reason),
      }    
  );
}
                       
class TrackingRepository { 
  void track(TrackEvent event) {
    _firebase.logEvent(
      name: event.name, 
      params: event.parameters,
    );    
    
    dev.log(event.name);
    // firebase
    // GA
    // tracker?.track...
  }
}
`

Reply:
> context.read<TrackingCubit>().trackEvent({
>       'name': 'pressed_button_delete',
>       'params': {
>         'event_type': 'button_action',
>         'user_type': 'admin',      
>         'user_id': 100,
>         'reason': 'special_limitation',
>         'additional_information': 12,
>       },
>     }
>   );

That's not what I was talking about, we're clearly on different pages here. Let me try and explain better.

We should aim for something like:
```
context.read<TrackingCubit>().trackEvent({
      name: 'pressed_button_delete', // these are so specific to each action that you'd have to analyze what marketing wants, to ensure there's a need to abstract this with the use of an enum or something. 
      eventType: EventType.buttonAction,
      userType: UserType.admin,
      userId: 100,
      reason: Reason.specialLimitation, // if it makes sense to predefine a set of reasons 
  });  
```
In my opinion actions are pretty different although they might share some "configuration". You're more familiar with the requirements and if you feel you want to wrap all this into an additional event class, go for it, but just make sure that there's a reusability utility to it.

Reply:
Reusability is definitely a benefit. For example consider tracking the same action on different entry points.